CMakeLists.txt

```diff
@@ -400,80 +400,88 @@ if(MSVC)
     set(CMAKE_DEBUG_POSTFIX "d")
     add_definitions(-D_CRT_SECURE_NO_DEPRECATE)
     add_definitions(-D_CRT_NONSTDC_NO_DEPRECATE)
+endif()
+
+# Check whether compiler supports SSE2 instrinics
+if(WITH_NATIVE_INSTRUCTIONS)
+    set(CMAKE_REQUIRED_FLAGS "${NATIVEFLAG}")
 else()
-    #
-    # Not MSVC, so we need to check if we have the MS-style SSE etc. intrinsics
-    #
-    if(WITH_NATIVE_INSTRUCTIONS)
-        set(CMAKE_REQUIRED_FLAGS "${NATIVEFLAG}")
-    else()
-        set(CMAKE_REQUIRED_FLAGS "${SSE2FLAG}")
-    endif()
+    set(CMAKE_REQUIRED_FLAGS "${SSE2FLAG}")
+endif()
+check_c_source_compile_or_run(
+    "#include <immintrin.h>
+    int main(void)
+    {
+        __m128i zero = _mm_setzero_si128();
+        (void)zero;
+        return 0;
+    }"
+    HAVE_SSE2_INTRIN
+)
+set(CMAKE_REQUIRED_FLAGS)
+
+# Check whether compiler supports SSE4 instrinics
+if(WITH_NATIVE_INSTRUCTIONS)
+    set(CMAKE_REQUIRED_FLAGS "${NATIVEFLAG}")
+else()
+    set(CMAKE_REQUIRED_FLAGS "${SSE4FLAG}")
+endif()
+check_c_source_compile_or_run(
+    "int main(void)
+    {
+        unsigned val = 0, h = 0;
+    #if defined(_MSC_VER)
+        { __asm mov edx, h __asm mov eax, val __asm crc32 eax, edx __asm mov val, eax }
+    #else
+        __asm__ __volatile__ ( \"crc32 %1,%0\" : \"+r\" (h) : \"r\" (val) );
```

Comment:
These are not compiler intrinsics, this is inline assembler code. Comment above should be changed, if this is actually what we want to check..?  I have not checked whether we use these inline assembler parts anywhere, but we generally try to avoid doing so because it makes the source harder to maintain and debug.
Also, the comment says we are testing for SSE4, but the asm references crc32 (part of SSE4.2), something the next check says it is testing.

So what does HAVE_SSE42_INTRIN really do (besides not checking for intrinsics at all, something HAVE_SSE42CRC_INTRIN does). I feel like the naming and comments need to get updated to match the actual code here.

I understand that this was also confusing before this commit, but this seems like a good time to fix it up while it is now being reworked.

HAVE_SSE42_INLINE_ASM would perhaps be a better name for this check.

Reply:
HAVE_SSE42CRC_INLINE_ASM would possibly be an even more relevant name, unless we use inline asm for something other than crc32 based on this test.

Reply:
As target, 32-bit Windows has inline assembler, 64-bit Windows does not have... 64-bit Windows uses MASM or intrinsics.

Reply:
I think the previous naming of HAVE_SSE42_INTRIN was a bit misleading and that HAVE_SSE42CRC_INLINE_ASM is more clear. I have made that change. I have also made a change to use inline asm for msvc in `insert_string_sse.c`.